cmd/quadlet/main.go

```diff
@@ -209,6 +210,37 @@ func enableServiceFile(outputPath string, service *parser.UnitFile) {
 	}
 }
 
+// warns if input is a "short name".  A "short name" refers to a
+// container image without a fully-qualified reference, and is hence
+// missing a registry (or domain).  Names including a digest are not
+// short names.
+//
+// Examples:
+//   - short names: "image:tag", "library/fedora"
+//   - not short names: "quay.io/image", "localhost/image:tag",
+//     "server.org:5000/lib/image", "image@sha256:...
+//
+// We implement a simple version of this from scratch here to avoid
+// a huge dependency in the generator just for a warning.
+func warnIfShortName(container *parser.UnitFile) {
+	imageName, ok := container.Lookup(quadlet.ContainerGroup, quadlet.KeyImage)
+	if !ok {
+		return
+	}
+	if strings.Contains(imageName, "@") {
```

Comment:
Use image/pkg/shortname IsShortName(imageName)

Reply:
Actually its not that bad, but it goes from 2.6BM to 3.1MB (stripped 1.8-2.21MB), and it pulls in things like containers/image/docker/reference with its 1.3 meg of regexps. 
Also adds these new deps: 
```
	libsubid.so.4 => /lib64/libsubid.so.4 (0x00007fb97a1ae000)
	libaudit.so.1 => /lib64/libaudit.so.1 (0x00007fb979fa8000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007fb979f7b000)
	libsemanage.so.2 => /lib64/libsemanage.so.2 (0x00007fb979f3a000)
	libcrypt.so.2 => /lib64/libcrypt.so.2 (0x00007fb979efe000)
	libacl.so.1 => /lib64/libacl.so.1 (0x00007fb979ef4000)
	libattr.so.1 => /lib64/libattr.so.1 (0x00007fb979eec000)
	libcap-ng.so.0 => /lib64/libcap-ng.so.0 (0x00007fb979ee2000)
	libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007fb979e45000)
	libsepol.so.2 => /lib64/libsepol.so.2 (0x00007fb979d8d000)
```

Reply:
Not to say any of this is horrible. But the generators are run during very early boot and block systemd from starting to read the unit files, so any slowness blocks the entire boot process.

Reply:
@vrothberg @mtrmac WDYT?

Reply:
I’d prefer using docker/reference but I’m not going to get into a huge fight about this…

Way too often, such handcrafted parsers are not actually correct (a confusion about the three uses of `:` is very common). That’s, AFAICS, not the case here, but it’s just a bad habit to get into. “Always parse the data using the provided API, never deal with raw strings” is easier to enforce with fewer counterexamples in the codebase.

From a quick check it seems that something like https://github.com/distribution/distribution/pull/3566 might bring us from 43 to 7 regexp compiles.

#16205 already talks about reducing the overhead of the reference parser even further. Would someone be willing to actually do that work.

Would one or two of the above be good enough?

---

The SELinux dependency is a bit surprising. How does that come about? Would moving (some variant of) `IsShortName` from `c/image/pkg/shortnames` to `c/image/docker/reference`, dropping things like `c/storage/pkg/homedir` (and the required rootless infrastructure), help?

---

The concept of short names is orthogonal to digested references. What is this function _actually_ warning about? I.e. “ideally use a fully-qualified name” does not motivate users to change things unless it says why they should do that extra work.

Reply:
> The SELinux dependency is a bit surprising. How does that come about? Would moving (some variant of) IsShortName from c/image/pkg/shortnames to c/image/docker/reference, dropping things like c/storage/pkg/homedir (and the required rootless infrastructure), help?

Yes, that would certainly help.

> The concept of short names is orthogonal to digested references. What is this function actually warning about? I.e. “ideally use a fully-qualified name” does not motivate users to change things unless it says why they should do that extra work.

User may well just ignore the warning.  I would go as far as throwing an error if we want to take the issue _really_ serious.

Reply:
I'm generally in favour of using dependencies with tested code, but the generator code runs during very special circumstances, and affect system-wide boot performace heavily, so in this particular case I'm against it.

In particular because this is really only a warning. The initial reason for recommending the use of fqn is that using a shortname on rhel unnecessary parses 300k of shorname aliases, which adds > 60mesec to boot. But it is also generally wise to use a fqn to make sure you're running exactly the image you expected.

Reply:
I don't think we should fail for a shortname, that seems unnecessarily harsh. However, giving a warning for something that is potentially quite slow at least gives users the possibility of acting on this.